frontend/src/customAxios.tsx

```diff
@@ -0,0 +1,82 @@
+import axios, { AxiosInstance, AxiosResponse } from "axios";
+import { isEmpty } from "lodash";
+import * as String from "constants/strings";
+import { store } from "App";
+
+const UNAUTHORIZED = 401;
+const MAINTENANCE = 503;
+const TEAPOT = 418;
+
+const defaultEnvelope = (x:any) => ({ data: { records: x } });
+
+const sleepRequest = (instance:AxiosInstance, milliseconds:number, originalRequest:any) => {
+  return new Promise((resolve) => {
+    setTimeout(() => resolve(instance(originalRequest)), milliseconds);
+  });
+};
+
+const errorCount:any = {};
+
+const CustomAxios = ({ errorToastMessage, selector, envelope = defaultEnvelope }: {errorToastMessage?: string; selector?:Function; envelope?: typeof defaultEnvelope} = {}) => {
+  const instance = axios.create();
+  instance.interceptors.request.use((config) => {
+    if (selector !== undefined) {
+      const state = store.getState();
+      const storedValue = selector(state);
+
+      if (!isEmpty(storedValue)) {
+        throw new axios.Cancel(JSON.stringify(envelope(storedValue)))
+      }
+    }
+    return config;
+  });
+
+  instance.interceptors.response.use(
+    (response) => response,
+    (error) => {
+      if (axios.isCancel(error)) {
+        return Promise.resolve(error.message);
+      }
+
+      errorCount[error.config.url] = errorCount[error.config.url]
+        ? errorCount[error.config.url] + 1
+        : 1;
+      const numberOfErrors = errorCount[error.config.url];
+
+      const status = error.response ? error.response.status : null;
+      const errorMessage =
+        errorToastMessage || (error.response && error.response.data.message) || String.ERROR;
+
+      if (status === UNAUTHORIZED || status === MAINTENANCE) {
+        window.location.reload(false);
+      } else if (status === TEAPOT || process.env.NODE_ENV === "development") {
+        // Do not retry in dev env, or for axios calls made in frontend tests, which are configured to return status code 418
+        //TODO: come up with a way of displaying request failures.
```

Comment:
admittedly this is a bit of a mess, but I don't have UI or requirements for:
1) some kind of progress bar or notification icon for async requests.
2) how to display async request failures.

Reply:
Ideally all requests would go through common library so that failures would be handled the same way.  Display a little floating div with a relevant, but generic error message when not in a Developer environment.